pkg/auth/auth_provider.go

```diff
@@ -55,6 +55,13 @@ func GetAuthProvider(config *common.Config) (*Provider, error) {
 			provider, err = NewAuthenticationToken(config.Token, defaultTransport)
 		case len(config.TokenFile) > 0:
 			provider, err = NewAuthenticationTokenFromFile(config.TokenFile, defaultTransport)
+		case len(config.KeyFile) > 0:
+			provider, err = NewAuthenticationOAuth2(
+				config.IssueEndpoint,
+				config.ClientID,
+				config.Audience,
+				config.KeyFile,
```

Comment:
The keyfile actually contains a clientid, since it represents an oauth2 client.  Better to think of the config as being +either+ a user grant or a service account grant.   User grant has `clientid`, `refresh token`, and `access token`.  Service account grant has `keyfile` which contains a client id and secret.

Meanwhile, the `audience` corresponds to the pulsar instance.